kbchat/kvstore.go

```diff
@@ -0,0 +1,218 @@
+	Error  Error                        `json:"error,omitempty"`
+}
+
+type ListNamespaces struct {
+	Result keybase1.KVListNamespaceResult `json:"result"`
+	Error  Error                          `json:"error,omitempty"`
+}
+
+type ListEntryKeys struct {
+	Result keybase1.KVListEntryResult `json:"result"`
+	Error  Error                      `json:"error,omitempty"`
+}
+
+type KVStoreAPI interface {
+	PutEntry(teamName string, namespace string, entryKey string, entryValue string, revision int) (keybase1.KVPutResult, error)
+	DeleteEntry(teamName string, namespace string, entryKey string, revision int) (keybase1.KVDeleteEntryResult, error)
+	GetEntry(teamName string, namespace string, entryKey string) (keybase1.KVGetResult, error)
+	ListNamespaces(teamName string) (keybase1.KVListNamespaceResult, error)
+	ListEntryKeys(teamName string, namespace string) (keybase1.KVListEntryResult, error)
+}
+
+func (a *API) PutEntry(teamName string, namespace string, entryKey string, entryValue string, revision int) (result keybase1.KVPutResult, err error) {
+
+	type PutArgs struct {
+		Method string `json:"method"`
+		Params struct {
+			Options struct {
+				Team       string `json:"team"`
+				Namespace  string `json:"namespace"`
+				EntryKey   string `json:"entryKey"`
+				EntryValue string `json:"entryValue"`
+				Revision   int    `json:"revision,omitempty"`
+			} `json:"options"`
+		} `json:"params"`
+	}
+
+	args := PutArgs{Method: "put"}
+	args.Params.Options.Team = teamName
+	args.Params.Options.Namespace = namespace
+	args.Params.Options.EntryKey = entryKey
+	args.Params.Options.EntryValue = entryValue
+
+	if revision != 0 {
+		args.Params.Options.Revision = revision
+	}
+
+	apiInput, err := json.Marshal(args)
+	if err != nil {
+		return result, err
+	}
+
+	cmd := a.runOpts.Command("kvstore", "api")
+	cmd.Stdin = strings.NewReader(string(apiInput))
+	bytes, err := cmd.Output()
+	if err != nil {
+		return result, fmt.Errorf("failed to call keybase kvstore api: %v", err)
+	}
+
+	entry := PutEntry{}
+	err = json.Unmarshal(bytes, &entry)
+	if err != nil {
+		return result, fmt.Errorf("failed to parse output from keybase kvstore api: %v", err)
+	}
+	if entry.Error.Message != "" {
+		return result, fmt.Errorf("received error from keybase kvstore api: %s", entry.Error.Message)
+	}
+	return entry.Result, nil
+}
+
+func (a *API) DeleteEntry(teamName string, namespace string, entryKey string, revision int) (result keybase1.KVDeleteEntryResult, err error) {
+
+	type DeleteArgs struct {
+		Method string `json:"method"`
+		Params struct {
+			Options struct {
+				Team      string `json:"team"`
+				Namespace string `json:"namespace"`
+				EntryKey  string `json:"entryKey"`
+				Revision  int    `json:"revision,omitempty"`
+			} `json:"options"`
+		} `json:"params"`
+	}
+
+	args := DeleteArgs{Method: "del"}
+	args.Params.Options.Team = teamName
+	args.Params.Options.Namespace = namespace
+	args.Params.Options.EntryKey = entryKey
```

Comment:
i'd probably do this like:
```go
args := DeleteArgs{
    Method: "del",
    Params: {
        Options: {
            Team: teamName,
            Namespace: namespace,
            EntryKey: entryKey,
        },
    },
}
```

Reply:
this suggestion didn't build for me; how i wrote it was the least verbose way i found to write it without getting `missing type in composite literal`.

Reply:
the other valid ways being: 
- not inlining the definitions for the nested structs `Params` and `Options`, or
- https://play.golang.org/p/vLJRxnuAKz